File: infera/router/breaker.py
###############################################################################
# Copyright (c) 2026, Advanced Micro Devices, Inc. All rights reserved.
#
# SPDX-License-Identifier: MIT
###############################################################################
"""Per-worker circuit breaker for routers that select their own target.

Failover alone is not enough. It retries a failed dispatch on another worker,
but the memory of that failure lives in a per-request ``tried`` set that is
discarded when the request returns -- so the next request scores the same broken
worker as if nothing happened, picks it again if cache locality says so, and
pays the failover cost again. A worker that is healthy to the platform and
broken for inference therefore taxes *every* request, indefinitely.

That worker is not hypothetical: it accepts the connection, answers ``/health``,
stays ``ACTIVE`` in discovery, and fails before the first byte. Kubernetes
cannot see it and neither can discovery. The router is the only component that
knows, and without this it forgets immediately.

Scope, deliberately narrow:

* **Only pre-first-byte failures trip it.** A failure after bytes have been
  streamed is already non-retryable by design -- the worker demonstrably served
  part of the request, and treating that as a health signal would open the
  breaker on ordinary client disconnects.
* **It never touches ``WorkerStatus``.** That field is owned by discovery; this
  is the router's private opinion, applied when filtering candidates.
* **Only routers that select use it.** ``direct.py`` has no failover because the
  gateway owns selection there, so a breaker would be wrong.

States are the usual three. ``closed`` routes normally. After
``failure_threshold`` consecutive failures the breaker goes ``open`` and the
worker is excluded for ``cooldown`` seconds. It then becomes ``half_open`` and
admits one probe at a time: success closes it and clears the count, failure
reopens it with the cooldown doubled, up to ``max_cooldown``. Backing off
matters because the common cause -- a worker wedged on a bad KV handoff -- does
not resolve on the first retry, and a fixed cooldown turns into a probe every
``cooldown`` seconds forever.

"One at a time" is bounded by ``probe_timeout`` rather than by waiting for an
outcome, because the outcome may never arrive: the slot is claimed while
filtering candidates, and only the one the policy dispatches to reports back.
"""

from __future__ import annotations

import logging
import time
from dataclasses import dataclass, field
from enum import Enum

from infera.server import metrics

logger = logging.getLogger(__name__)

_STATE_VALUE = {"closed": 0, "half_open": 1, "open": 2}


def _observe(worker_id: str, state) -> None:
    """Mirror a state change to Prometheus. Never allowed to fail a request --
    an unregistered collector or a duplicate registry must not take out the
    data plane."""
    try:
        metrics.worker_breaker_state.labels(worker_id=worker_id).set(_STATE_VALUE[state.value])
    except Exception:  # pragma: no cover
        pass


def is_worker_fault(status: int) -> bool:
    """True if an HTTP status is evidence about the *worker*, not the request.

    Failover retries on any pre-first-byte error, including 4xx -- that is
    correct, since a 400 costs nothing to re-ask. Feeding 4xx to the breaker is
    not: a malformed request returns 400 from every worker it touches, so the
    breaker would trip the entire healthy fleet on one bad client. 429 is
    excluded for a different reason -- it means "full right now", which the
    policy's load accounting already routes around, and a 5s cooldown with
    doubling is far too heavy a response to transient backpressure.
    """
    return status >= 500 or status == 0


class BreakerState(str, Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


@dataclass
class _Entry:
    consecutive_failures: int = 0
    state: BreakerState = BreakerState.CLOSED
    # Wall time after which an open breaker becomes half-open.
    opens_until: float = 0.0
    # Cooldown applied on the *next* trip; doubles each time a probe fails.
    next_cooldown: float = 0.0
    # When the outstanding half-open probe was admitted, so only one is in
    # flight at a time. None means the slot is free.
    probe_started_at: float | None = None
    trips: int = 0


@dataclass
class CircuitBreaker:
    """Tracks per-worker health as seen by dispatch outcomes.

    Not thread-safe by design: routers drive it from a single asyncio loop, and
    a lock here would sit on the hot path of every request for no benefit.
    """

    failure_threshold: int = 3
    cooldown: float = 5.0
    max_cooldown: float = 60.0
    #: How long a claimed probe slot is honoured before it is reclaimed.
    #:
    #: Claiming and releasing the slot are not paired: ``filter`` claims one for
    #: every candidate it lets through, and the policy then dispatches exactly
    #: one of them, so the rest are never told how they did. A 4xx, a client
    #: disconnect or a request that never returns leaves the slot held too.
    #: Bounding the claim keeps any of those from wedging a healthy worker out
    #: of rotation permanently. The cost of reclaiming too early is one extra
    #: probe; the cost of never reclaiming is a worker lost until restart.
    probe_timeout: float = 60.0
    #: Injectable clock, so tests do not sleep.
    now: object = field(default=time.monotonic)
    _entries: dict[str, _Entry] = field(default_factory=dict, init=False)

    # --- queries ------------------------------------------------------------

    def _entry(self, worker_id: str) -> _Entry:
        e = self._entries.get(worker_id)
        if e is None:
            e = _Entry(next_cooldown=self.cooldown)
            self._entries[worker_id] = e
        return e

    @property
    def enabled(self) -> bool:
        """A threshold of 0 or less turns the breaker off entirely, so an
        operator can fall back to plain failover without a code change."""
        return self.failure_threshold > 0

    def allows(self, worker_id: str) -> bool:
        """True if this worker may be dispatched to right now.

        Transitions open -> half_open as a side effect when the cooldown has
        elapsed, because the alternative is a separate timer whose only job is
        to flip a flag that this function already has to check.
        """
        if not self.enabled:
            return True
        e = self._entries.get(worker_id)
        if e is None or e.state is BreakerState.CLOSED:
            return True
        now = self.now()
        if e.state is BreakerState.OPEN:
            if now < e.opens_until:
                return False
            e.state = BreakerState.HALF_OPEN
            e.probe_started_at = None
            _observe(worker_id, e.state)
            logger.info("breaker: worker %s half-open, admitting one probe", worker_id)
        # half_open: one probe at a time, and only for as long as a probe could
        # plausibly still be running -- see probe_timeout for why the claim has
        # to expire rather than wait for an outcome that may never come.
        if e.probe_started_at is not None:
            if now - e.probe_started_at < self.probe_timeout:
                return False
            logger.info(
                "breaker: worker %s probe slot unclaimed after %.0fs; admitting another",
                worker_id,
                self.probe_timeout,
            )
        e.probe_started_at = now
        return True

    def filter(self, workers):
        """Drop workers whose breaker is open. Returns a list.

        If every candidate is open, returns them all rather than nothing: a
        request served by a probably-bad worker beats a 503 when there is no
        alternative, and refusing to route would turn a partial outage into a
        total one.
        """
        allowed = [w for w in workers if self.allows(self._id_of(w))]
        if allowed:
            return allowed
        if workers:
            logger.warning(
                "breaker: all %d candidate(s) open; routing anyway rather than failing",
                len(workers),
            )
        return list(workers)

    @staticmethod
    def _id_of(w) -> str:
        # Accepts a WorkerInfo or anything exposing .worker_id.
        return getattr(w, "worker_id", None) or str(w)

    def state_of(self, worker_id: str) -> BreakerState:
        e = self._entries.get(worker_id)
        return e.state if e else BreakerState.CLOSED

    # --- outcomes -----------------------------------------------------------

    def record_success(self, worker_id: str) -> None:
        e = self._entries.get(worker_id)
        if e is None:
            return
        if e.state is not BreakerState.CLOSED:
            logger.info("breaker: worker %s recovered, closing", worker_id)
        e.consecutive_failures = 0
        e.state = BreakerState.CLOSED
        e.probe_started_at = None
        e.next_cooldown = self.cooldown
        _observe(worker_id, e.state)

    def record_neutral(self, worker_id: str) -> None:
        """Release the probe slot without scoring the worker either way.

        For an outcome that says nothing about worker health -- a 4xx, which
        every worker would answer identically, or a 429, which is backpressure
        the policy already routes around. Counting it as recovery is as wrong
        as counting it as failure: it would reset the failure count and close
        an open breaker, so a worker alternating 500s and 400s would never
        accumulate the consecutive failures needed to trip. But the slot such a
        request consumed must still come back, or one bad client can wedge a
        recovering worker out of rotation.
        """
        e = self._entries.get(worker_id)
        if e is None:
            return
        e.probe_started_at = None

    def forget(self, worker_id: str) -> None:
        """Drop everything remembered about a worker that has left the fleet.

        Worker ids are addresses and a rebuilt Pod never reuses one, so without
        this every rollout strands another entry -- and another pair of
        Prometheus series, since both are labelled by worker id.
        """
        if self._entries.pop(worker_id, None) is None:
            return
        for collector in (metrics.worker_breaker_state, metrics.worker_breaker_trips_total):
            try:
                collector.remove(worker_id)
            except Exception:  # noqa: BLE001 - never registered, or already gone
                pass

    def record_failure(self, worker_id: str) -> None:
        """Record a pre-first-byte dispatch failure."""
        if not self.enabled:
            return
        e = self._entry(worker_id)
        e.consecutive_failures += 1
        was_probe = e.state is BreakerState.HALF_OPEN
        e.probe_started_at = None

        if was_probe:
            # A failed probe reopens immediately and backs off further, without
            # waiting for the threshold again -- we already know it is bad.
            e.next_cooldown = min(e.next_cooldown * 2, self.max_cooldown)
            self._open(worker_id, e)
            return
        if e.consecutive_failures >= self.failure_threshold:
            self._open(worker_id, e)

    def _open(self, worker_id: str, e: _Entry) -> None:
        # A trip is an edge into exclusion, not every failure that lands while
        # the worker is already excluded. A failed probe counts: it is a fresh
        # verdict on a worker that was given another chance, and the doubling
        # cooldown bounds how often one can happen. A failure while already
        # open does not -- those arrive at the request rate, via the all-open
        # fallback, and counting them turns the metric into a request counter
        # that drowns out real trips and prints the warning on every request.
        newly_tripped = e.state is not BreakerState.OPEN
        e.state = BreakerState.OPEN
        e.opens_until = self.now() + e.next_cooldown
        _observe(worker_id, e.state)
        if not newly_tripped:
            return
        e.trips += 1
        try:
            metrics.worker_breaker_trips_total.labels(worker_id=worker_id).inc()
        except Exception:  # pragma: no cover - metrics must never break routing
            pass
        logger.warning(
            "breaker: worker %s open for %.1fs after %d consecutive failure(s)",
            worker_id,
            e.next_cooldown,
            e.consecutive_failures,
        )

    # --- introspection for metrics / tests -----------------------------------

    def snapshot(self) -> dict[str, dict]:
        return {
            wid: {
                "state": e.state.value,
                "consecutive_failures": e.consecutive_failures,
                "trips": e.trips,
            }
            for wid, e in self._entries.items()
        }

File: rust/router/src/breaker.rs
///////////////////////////////////////////////////////////////////////////////
// Copyright (c) 2026, Advanced Micro Devices, Inc. All rights reserved.
//
// SPDX-License-Identifier: MIT
///////////////////////////////////////////////////////////////////////////////
//! Per-worker circuit breaker. Mirrors `infera/router/breaker.py` — same three
//! states, same thresholds, same all-open fallback — so the two data planes
//! behave identically under a wedged worker.
//!
//! Failover on its own is not enough. It retries a failed dispatch elsewhere,
//! but the memory of that failure lives in a per-request `tried` set that is
//! dropped when the request returns, so the next request scores the same broken
//! worker as if nothing had happened. A worker that answers `/health`, stays
//! ACTIVE in etcd, and fails before the first byte therefore taxes *every*
//! request, indefinitely.
//!
//! Unlike the Python side this is shared across tokio worker threads, so the
//! map lives behind a `Mutex`. The critical sections are a hash lookup and a
//! few integer writes; contention is not a concern at any plausible request
//! rate, and a lock-free design here would buy nothing for the complexity.

use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, Instant};

/// Statuses that are evidence about the *worker*, not the request.
///
/// Failover retries on any pre-first-byte error including 4xx, which is
/// correct — re-asking costs nothing. Feeding 4xx to the breaker is not: a
/// malformed request returns 400 from every worker it touches, so one bad
/// client would trip the whole healthy fleet. 429 is excluded for a different
/// reason: it means "full right now", which load accounting already routes
/// around, and a doubling cooldown is far too heavy for transient backpressure.
pub fn is_worker_fault(status: u16) -> bool {
    status >= 500 || status == 0
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BreakerState {
    Closed,
    Open,
    HalfOpen,
}

impl BreakerState {
    pub fn as_str(self) -> &'static str {
        match self {
            BreakerState::Closed => "closed",
            BreakerState::Open => "open",
            BreakerState::HalfOpen => "half_open",
        }
    }
}

#[derive(Debug)]
struct Entry {
    consecutive_failures: u32,
    state: BreakerState,
    /// Instant after which an open breaker becomes half-open.
    opens_until: Instant,
    /// Cooldown applied on the *next* trip; doubles each time a probe fails.
    next_cooldown: Duration,
    /// Set while a half-open probe is in flight, so only one is admitted.
    probe_in_flight: bool,
    trips: u64,
}

pub struct CircuitBreaker {
    failure_threshold: u32,
    cooldown: Duration,
    max_cooldown: Duration,
    entries: Mutex<HashMap<String, Entry>>,
}

impl CircuitBreaker {
    pub fn new(failure_threshold: u32, cooldown: Duration, max_cooldown: Duration) -> Self {
        Self {
            failure_threshold,
            cooldown,
            max_cooldown,
            entries: Mutex::new(HashMap::new()),
        }
    }

    /// Whether this worker may be dispatched to right now.
    ///
    /// Transitions Open -> HalfOpen as a side effect once the cooldown has
    /// elapsed, because the alternative is a background timer whose only job is
    /// to flip a flag this function already has to read. Call it once per
    /// candidate per request: in HalfOpen it *consumes* the single probe slot.
    pub fn allows(&self, worker_id: &str) -> bool {
        self.allows_at(worker_id, Instant::now())
    }

    /// A threshold of 0 turns the breaker off entirely, so an operator can fall
    /// back to plain failover without a code change.
    fn enabled(&self) -> bool {
        self.failure_threshold > 0
    }

    fn allows_at(&self, worker_id: &str, now: Instant) -> bool {
        if !self.enabled() {
            return true;
        }
        let mut map = self.entries.lock().expect("breaker mutex poisoned");
        let Some(e) = map.get_mut(worker_id) else {
            return true;
        };
        match e.state {
            BreakerState::Closed => return true,
            BreakerState::Open => {
                if now < e.opens_until {
                    return false;
                }
                e.state = BreakerState::HalfOpen;
                e.probe_in_flight = false;
                tracing::info!(worker = worker_id, "breaker half-open, admitting one probe");
            }
            BreakerState::HalfOpen => {}
        }
        if e.probe_in_flight {
            return false;
        }
        e.probe_in_flight = true;
        true
    }

    /// Drop workers whose breaker is open.
    ///
    /// If *every* candidate is open the full list is returned instead of an
    /// empty one: a request served by a probably-bad worker beats a guaranteed
    /// 503, and refusing to route would turn a partial outage into a total one.
    pub fn filter<W: Clone>(&self, workers: &[W], id_of: impl Fn(&W) -> &str) -> Vec<W> {
        let allowed: Vec<W> = workers
            .iter()
            .filter(|w| self.allows(id_of(w)))
            .cloned()
            .collect();
        if !allowed.is_empty() {
            return allowed;
        }
        if !workers.is_empty() {
            tracing::warn!(
                candidates = workers.len(),
                "breaker: all candidates open; routing anyway rather than failing"
            );
        }
        workers.to_vec()
    }

    pub fn record_success(&self, worker_id: &str) {
        let mut map = self.entries.lock().expect("breaker mutex poisoned");
        if let Some(e) = map.get_mut(worker_id) {
            if e.state != BreakerState::Closed {
                tracing::info!(worker = worker_id, "breaker: worker recovered, closing");
            }
            e.consecutive_failures = 0;
            e.state = BreakerState::Closed;
            e.probe_in_flight = false;
            e.next_cooldown = self.cooldown;
        }
    }

    /// Record a pre-first-byte dispatch failure. Callers must gate this on
    /// [`is_worker_fault`] when the failure carries an HTTP status.
    pub fn record_failure(&self, worker_id: &str) {
        self.record_failure_at(worker_id, Instant::now());
    }

    fn record_failure_at(&self, worker_id: &str, now: Instant) {
        if !self.enabled() {
            return;
        }
        let mut map = self.entries.lock().expect("breaker mutex poisoned");
        let e = map.entry(worker_id.to_string()).or_insert_with(|| Entry {
            consecutive_failures: 0,
            state: BreakerState::Closed,
            opens_until: now,
            next_cooldown: self.cooldown,
            probe_in_flight: false,
            trips: 0,
        });
        e.consecutive_failures += 1;
        let was_probe = e.state == BreakerState::HalfOpen;
        e.probe_in_flight = false;

        if was_probe {
            // A failed probe reopens immediately and backs off further, without
            // waiting out the threshold again — we already know it is bad. The
            // common cause (a worker wedged on a bad KV handoff) does not clear
            // on the first retry, and a fixed cooldown would probe it at a
            // constant rate forever.
            e.next_cooldown = (e.next_cooldown * 2).min(self.max_cooldown);
        } else if e.consecutive_failures < self.failure_threshold {
            return;
        }
        e.state = BreakerState::Open;
        e.opens_until = now + e.next_cooldown;
        e.trips += 1;
        tracing::warn!(
            worker = worker_id,
            cooldown_s = e.next_cooldown.as_secs_f64(),
            failures = e.consecutive_failures,
            "breaker: worker open"
        );
    }

    pub fn state_of(&self, worker_id: &str) -> BreakerState {
        self.entries
            .lock()
            .expect("breaker mutex poisoned")
            .get(worker_id)
            .map(|e| e.state)
            .unwrap_or(BreakerState::Closed)
    }

    /// `(worker_id, state, trips)` for metrics export.
    pub fn snapshot(&self) -> Vec<(String, BreakerState, u64)> {
        let map = self.entries.lock().expect("breaker mutex poisoned");
        let mut out: Vec<_> = map
            .iter()
            .map(|(k, e)| (k.clone(), e.state, e.trips))
            .collect();
        out.sort_by(|a, b| a.0.cmp(&b.0));
        out
    }
}

impl Default for CircuitBreaker {
    fn default() -> Self {
        Self::new(3, Duration::from_secs(5), Duration::from_secs(60))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Time is driven forward explicitly rather than slept, so the cooldown
    /// behaviour is tested at full speed and deterministically.
    fn cb() -> CircuitBreaker {
        CircuitBreaker::new(3, Duration::from_secs(5), Duration::from_secs(20))
    }

    #[derive(Clone)]
    struct W(&'static str);

    #[test]
    fn threshold_zero_disables_it() {
        // Without the guard, `failures >= 0` would trip on the first failure —
        // the exact opposite of what --breaker-failure-threshold=0 promises.
        let b = CircuitBreaker::new(0, Duration::from_secs(5), Duration::from_secs(20));
        for _ in 0..20 {
            b.record_failure("w1");
        }
        assert!(b.allows("w1"));
        assert_eq!(b.state_of("w1"), BreakerState::Closed);
        let ws = vec![W("a"), W("b")];
        assert_eq!(b.filter(&ws, |w| w.0).len(), 2);
    }

    #[test]
    fn unknown_worker_is_allowed() {
        let b = cb();
        assert!(b.allows("w1"));
        assert_eq!(b.state_of("w1"), BreakerState::Closed);
    }

    #[test]
    fn failures_below_threshold_do_not_open() {
        let b = cb();
        b.record_failure("w1");
        b.record_failure("w1");
        assert!(b.allows("w1"));
        assert_eq!(b.state_of("w1"), BreakerState::Closed);
    }

    #[test]
    fn opens_at_threshold_and_excludes() {
        let b = cb();
        for _ in 0..3 {
            b.record_failure("w1");
        }
        assert_eq!(b.state_of("w1"), BreakerState::Open);
        assert!(!b.allows("w1"));
    }

    #[test]
    fn success_resets_the_count() {
        let b = cb();
        b.record_failure("w1");
        b.record_failure("w1");
        b.record_success("w1");
        b.record_failure("w1");
        b.record_failure("w1");
        assert_eq!(b.state_of("w1"), BreakerState::Closed);
    }

    #[test]
    fn half_open_admits_exactly_one_probe() {
        let b = cb();
        let t0 = Instant::now();
        for _ in 0..3 {
            b.record_failure_at("w1", t0);
        }
        let t1 = t0 + Duration::from_millis(5_100);
        assert!(b.allows_at("w1", t1), "cooldown elapsed -> one probe");
        assert_eq!(b.state_of("w1"), BreakerState::HalfOpen);
        assert!(
            !b.allows_at("w1", t1),
            "a second concurrent request must not also probe"
        );
    }

    #[test]
    fn successful_probe_closes() {
        let b = cb();
        let t0 = Instant::now();
        for _ in 0..3 {
            b.record_failure_at("w1", t0);
        }
        let t1 = t0 + Duration::from_millis(5_100);
        b.allows_at("w1", t1);
        b.record_success("w1");
        assert_eq!(b.state_of("w1"), BreakerState::Closed);
        assert!(b.allows_at("w1", t1));
    }

    #[test]
    fn failed_probe_reopens_with_doubled_cooldown() {
        let b = cb();
        let t0 = Instant::now();
        for _ in 0..3 {
            b.record_failure_at("w1", t0);
        }
        let t1 = t0 + Duration::from_millis(5_100);
        b.allows_at("w1", t1);
        b.record_failure_at("w1", t1); // probe fails -> reopen, 5s -> 10s
        assert_eq!(b.state_of("w1"), BreakerState::Open);

        let t2 = t1 + Duration::from_millis(5_100); // old cooldown would be up
        assert!(!b.allows_at("w1", t2), "backoff must have doubled");
        let t3 = t1 + Duration::from_millis(10_100);
        assert!(b.allows_at("w1", t3));
    }

    #[test]
    fn cooldown_is_capped() {
        let b = cb();
        let mut t = Instant::now();
        for _ in 0..3 {
            b.record_failure_at("w1", t);
        }
        for _ in 0..6 {
            t += Duration::from_secs(1000);
            b.allows_at("w1", t);
            b.record_failure_at("w1", t);
        }
        assert!(
            b.allows_at("w1", t + Duration::from_millis(20_100)),
            "cooldown must not grow without bound"
        );
    }

    #[test]
    fn filter_drops_open_workers() {
        let b = cb();
        for _ in 0..3 {
            b.record_failure("bad");
        }
        let ws = vec![W("good"), W("bad")];
        let got = b.filter(&ws, |w| w.0);
        assert_eq!(got.len(), 1);
        assert_eq!(got[0].0, "good");
    }

    #[test]
    fn filter_returns_all_when_every_worker_is_open() {
        let b = cb();
        for id in ["a", "b"] {
            for _ in 0..3 {
                b.record_failure(id);
            }
        }
        let ws = vec![W("a"), W("b")];
        assert_eq!(b.filter(&ws, |w| w.0).len(), 2);
    }

    #[test]
    fn filter_of_empty_is_empty() {
        let b = cb();
        let ws: Vec<W> = vec![];
        assert!(b.filter(&ws, |w| w.0).is_empty());
    }

    #[test]
    fn workers_are_independent() {
        let b = cb();
        for _ in 0..3 {
            b.record_failure("bad");
        }
        assert!(b.allows("good"));
    }

    #[test]
    fn success_on_unknown_worker_is_harmless() {
        let b = cb();
        b.record_success("never-seen");
        assert!(b.allows("never-seen"));
    }

    #[test]
    fn snapshot_reports_trips() {
        let b = cb();
        let t0 = Instant::now();
        for _ in 0..3 {
            b.record_failure_at("w1", t0);
        }
        let t1 = t0 + Duration::from_millis(5_100);
        b.allows_at("w1", t1);
        b.record_failure_at("w1", t1);
        let snap = b.snapshot();
        assert_eq!(snap.len(), 1);
        assert_eq!(snap[0].1, BreakerState::Open);
        assert_eq!(snap[0].2, 2, "initial trip plus the failed probe");
    }

    #[test]
    fn client_errors_are_not_worker_faults() {
        for s in [400u16, 404, 422, 429] {
            assert!(!is_worker_fault(s), "{s} must not trip the breaker");
        }
        for s in [0u16, 500, 502, 503, 504] {
            assert!(is_worker_fault(s), "{s} must trip the breaker");
        }
    }

    #[test]
    fn a_bad_client_cannot_trip_the_fleet() {
        let b = cb();
        let ws = vec![W("a"), W("b"), W("c")];
        for _ in 0..10 {
            for w in b.filter(&ws, |w| w.0) {
                if is_worker_fault(400) {
                    b.record_failure(w.0);
                }
            }
        }
        for w in &ws {
            assert_eq!(b.state_of(w.0), BreakerState::Closed);
        }
    }

    #[test]
    fn the_regression_this_exists_for() {
        // A worker that fails every dispatch must stop being selected. Before
        // this type existed `tried` was per-request, so `bad` was offered on
        // all ten requests.
        let b = cb();
        let ws = vec![W("good"), W("bad")];
        let mut offered_bad = 0;
        for _ in 0..10 {
            let cands = b.filter(&ws, |w| w.0);
            if cands.iter().any(|w| w.0 == "bad") {
                offered_bad += 1;
                b.record_failure("bad");
            }
            b.record_success("good");
        }
        assert_eq!(offered_bad, 3, "bad worker must stop being offered");
    }

    #[test]
    fn concurrent_probes_admit_only_one() {
        // The Python breaker is single-loop; this one is shared across tokio
        // threads, so the half-open slot has to be safe under real contention.
        use std::sync::Arc;
        let b = Arc::new(cb());
        for _ in 0..3 {
            b.record_failure("w1");
        }
        // Force half-open by driving the clock through the private hook.
        let t = Instant::now() + Duration::from_secs(6);
        assert!(b.allows_at("w1", t));
        b.record_failure_at("w1", t); // back to open, then reopen at 10s
        let t2 = t + Duration::from_secs(11);

        let admitted = Arc::new(std::sync::atomic::AtomicUsize::new(0));
        let hs: Vec<_> = (0..16)
            .map(|_| {
                let b = b.clone();
                let admitted = admitted.clone();
                std::thread::spawn(move || {
                    if b.allows_at("w1", t2) {
                        admitted.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
                    }
                })
            })
            .collect();
        for h in hs {
            h.join().unwrap();
        }
        assert_eq!(
            admitted.load(std::sync::atomic::Ordering::SeqCst),
            1,
            "exactly one of 16 racing threads may probe"
        );
    }
}
